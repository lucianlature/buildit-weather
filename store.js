import { createStore, compose, applyMiddleware } from 'redux'
import thunkMiddleware from 'redux-thunk'
import fetch from 'isomorphic-fetch'

const API_KEY = '060c011784ac004d3668a0696a4068c5'

const initialState = {
  city: {
      name: 'Woking',
      country: 'UK'
  },
  forecast: []
}

export const requestForecast = location => ({
  type: actionTypes.REQUEST_FORECAST,
  location
})

function receiveForecast(location, json) {
  return {
    type: actionTypes.RECEIVE_FORECAST,
    location,
    forecast: json.list,
    receivedAt: Date.now()
  }
}

function shouldFetchForecast(state, location) {
  const forecast = state.forecast
  return !forecast.length
}

export const actionTypes = {
  REQUEST_FORECAST: 'REQUEST_FORECAST',
  RECEIVE_FORECAST: 'RECEIVE_FORECAST',
  CHANGE_LOCATION: 'CHANGE_LOCATION',
  EXCHANGE: 'EXCHANGE',
}

// REDUCER
export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.RECEIVE_FORECAST:
      return Object.assign({}, state, { forecast: action.forecast })
    case actionTypes.CHANGE_LOCATION:
      return Object.assign({}, state, { city: action.city })
    default: return state
  }
}

// ACTIONS
export const serverLoadLocation = (isServer) => dispatch => {
  return dispatch({ type: actionTypes.LOAD_LOCATION })
}

export const fetchForecastIfNeeded = (location) => (dispatch, getState) => {
  // Note that the function also receives getState()
  // which lets you choose what to dispatch next.

  // This is useful for avoiding a network request if
  // a cached value is already available.
  if (shouldFetchForecast(getState(), location)) {
    // Dispatch a thunk from thunk!
    return dispatch(fetchForecast(location))
  } else {
    // Let the calling code know there's nothing to wait for.
    return Promise.resolve()
  }
}

export const fetchForecast = location => dispatch => {
  dispatch(requestForecast(location))

  return fetch(`http://api.openweathermap.org/data/2.5/forecast?APPID=${API_KEY}&units=metric&q=${location}`)
    .then(response => response.json())
    .then(json => dispatch(receiveForecast(location, json)))
}

export const changeLocation = (event) => dispatch => {
  let location = String(event.target.value)

  if (location.length < 3) {
      return
  }

  const [ name, country ] = location.split(',')

  if (!name) {
      return
  }

  if (!country) {
      return
  }

  dispatch({type: actionTypes.CHANGE_LOCATION, city: { name, country }})
  return dispatch(fetchForecast(location))
}

// Get the Redux DevTools extension and fallback to a no-op function
let devtools = f => f

export const initStore = (initialState = initialState) => {
  if (process.browser) {
    if (window.__REDUX_DEVTOOLS_EXTENSION__) {
      devtools = window.__REDUX_DEVTOOLS_EXTENSION__()
    }
  }
  return createStore(reducer, initialState, compose(applyMiddleware(thunkMiddleware), devtools))
}
