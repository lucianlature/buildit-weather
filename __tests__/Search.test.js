
import { shallow, mount } from 'enzyme'
import React from 'react'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'

import ConnectedSearch, { Search } from '../components/Search'
import ConnectedBuilditPage, { BuilditPage } from '../pages/index'
import { fetchForecast, changeLocation, reducer } from '../store'

// Snapshot for IndexPage React Component
describe('>>> SEARCH --- Snapshot', () => {
    it('+++ capturing Snapshot of Search', () => {
        const city = { name: 'Woking', country: 'GB' }

        const renderedValue =  renderer.create(<Search city={city} />).toJSON()
        expect(renderedValue).toMatchSnapshot()
    })
})

//*******************************************************************************************************
describe('>>> SEARCH --- Shallow Render REACT COMPONENTS', () => {
  let wrapper
  const city = { name: 'Woking', country: 'GB' }

  beforeEach(()=>{
    wrapper = shallow(<Search city={city} />)    
  })

  it('+++ render the DUMB component', () => {
    expect(wrapper.length).toEqual(1)
  })

  it('+++ contains form', () => {
    expect(wrapper.find('form').length).toEqual(1)
  })
})

//*******************************************************************************************************
describe('>>> SEARCH --- REACT-REDUX (Shallow + passing the {store} directly)', () => {
    const initialState = {
        city: {
            name: 'Woking',
            country: 'UK'
        },
        forecast: []
    }
    const mockStore = configureStore()
    let store, container

    beforeEach(() => {
        store = mockStore(initialState)
        container = shallow(<ConnectedSearch store={store} /> ) 
    })

    it('+++ render the connected(SMART) component', () => {
       expect(container.length).toEqual(1)
    })

    it('+++ check Prop matches with initialState', () => {
       expect(container.prop('city')).toEqual(initialState.city)
    })
})

//*******************************************************************************************************
describe('>>> SEARCH --- REACT-REDUX (Mount + wrapping in <Provider>)', () => {
    const initialState = {
        city: {
            name: 'Woking',
            country: 'UK'
        },
        forecast: []
    }
    const mockStore = configureStore([thunk])
    let store, wrapper

    beforeEach(() => {
        store = mockStore(initialState)
        wrapper = mount( <ConnectedBuilditPage store={store}><ConnectedSearch /></ConnectedBuilditPage> )
    })

    it('+++ contains form', () => {
        expect(wrapper.find('form').length).toEqual(1)
    })

    it('+++ render the connected(SMART) component', () => {
       expect(wrapper.find(ConnectedSearch).length).toEqual(1)
    })

    it('+++ check Prop matches with initialState', () => {
       expect(wrapper.find(Search).prop('city')).toEqual(initialState.city)
    })

    it('+++ check action on dispatching ', () => {
        let action
        
        store.dispatch(fetchForecast({ location:  'London,UK' }))
        
        action = store.getActions()

        expect(action[0].type).toBe('REQUEST_FORECAST')
    })
})

//*******************************************************************************************************
describe('>>> SEARCH --- REACT-REDUX (actual Store + reducers) more of Integration Testing', () => {
    const initialState = {
        city: {
            name: 'Woking',
            country: 'UK'
        },
        forecast: []
    }
    let store, wrapper

    beforeEach(() => {
        store = createStore(reducer, initialState, applyMiddleware(thunk))
        wrapper = mount( <ConnectedBuilditPage store={store}><ConnectedSearch /></ConnectedBuilditPage> )
    })

    it('+++ check Prop matches with initialState', () => {
        store.dispatch(changeLocation({ target: { value:  'London,UK' } }))
        expect(wrapper.find(Search).prop('city')).toEqual({"country": "UK", "name": "London"})
    });

})