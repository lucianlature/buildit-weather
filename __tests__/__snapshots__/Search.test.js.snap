// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> SEARCH --- Snapshot +++ capturing Snapshot of Search 1`] = `
<section
  className="search"
  data-jsx={2278698433}
>
  <form
    data-jsx={2278698433}
    id="search"
    method="get"
  >
    <input
      data-jsx={2278698433}
      defaultValue="Woking,GB"
      name="q"
      onChange={[Function]}
      placeholder="Enter city name..."
      size="40"
      type="text"
    />
  </form>
</section>
`;
